gapir/cc/in_memory_resource_cache.cpp

```diff
@@ -23,174 +23,201 @@
-    size_t offset = (mHead->offset + resource.size) % mBufferSize;
-    auto next = new Block(offset, space);
-    next->linkAfter(mHead);
-    mHead->size = resource.size;
+  mResourceIndex.clear();
+  mResources.clear();
+
+  mMemoryUse = 0;
+}
+
+std::map<unsigned int, std::pair<Resource, std::shared_ptr<char> > >::iterator
+InMemoryResourceCache::findCache(const Resource& res) {
+  auto resIndex = mResourceIndex.find(res.getID());
+  if (resIndex == mResourceIndex.end()) {
+    return mResources.end();
   }
 
-  // Update mCache.
-  mCache.erase(mHead->id);
-  mCache.emplace(resource.id, mHead->offset);
-  mHead->id = resource.id;
+  auto resRecord = mResources.find(resIndex->second);
+  return resRecord;
+}
+
+bool InMemoryResourceCache::evictLeastRecentlyUsed() {
+  auto lru = mResources.begin();
+  if (lru != mResources.end()) {
+    lru->second.second = nullptr;
+    mMemoryUse -= lru->second.first.getSize();
 
-  // Copy data.
-  if (mHead->offset + resource.size <= mBufferSize) {
-    memcpy(mBuffer - mHead->offset - resource.size, data, resource.size);
+    mResourceIndex.erase(lru->second.first.getID());
+    mResources.erase(lru);
   } else {
-    // Wraps the end of the buffer
-    const uint8_t* dst = reinterpret_cast<const uint8_t*>(data);
-    size_t a = mBufferSize - mHead->offset;
-    size_t b = resource.size - a;
-    memcpy(mBuffer - mBufferSize, dst, a);
-    memcpy(mBuffer - b, dst + a, b);
+    return false;
   }
 
-  // Move head on to the next block.
-  mHead = mHead->next;
   return true;
 }
 
-bool InMemoryResourceCache::hasCache(const Resource& resource) {
-  return mCache.find(resource.id) != mCache.end();
-}
-
-bool InMemoryResourceCache::loadCache(const Resource& resource, void* data) {
-  if (!hasCache(resource)) {
+bool InMemoryResourceCache::loadCacheMiss(const Resource& res, void* target) {
+  // How much could we prefetch if we wanted to completely fill (100% eviction
+  // rate) the cache?
+  size_t possiblePrefetch =
+      res.getSize() < totalCacheSize() ? (totalCacheSize() - res.getSize()) : 0;
+  // Lets prefetch 10% of that maximum figure.
+  size_t prefetch = possiblePrefetch / 10;
+
+  // Try to anticipate the next few resources.
+  std::vector<Resource> anticipated = anticipateNextResources(res, prefetch);
+
+  // Don't forget the resource that kicked this cache miss off.
+  anticipated.push_back(res);
+
+  // Prefetch the anticipated resources first, so if anything gets evicted it's
+  // not the resource that initiated the miss. Then finally fetch the resource
+  // that caused the cache miss. This is last in the vector so it will not be
+  // kicked by the LRU Policy.
+  prefetchImpl(&anticipated[0], anticipated.size(), true);
+
+  // Unless something went very wrong, the data should now be in cache.
+  auto resRecord = findCache(res);
+  if (resRecord == mResources.end()) {
+    GAPID_ERROR(
+        "Cache miss prefetch failed for resource %s. This is probably very "
```

Comment:
I didn't want to kill the program, because depending on how the cache is being used, the program may actually be able to recover (See CachedResourceLoader's use of ResourceCache).